### ISSUE TITLE
Command and control for Lora adapter

### ISSUE THREAD

**Issue by BobClaerhout** [eclipse/hono#2482]
Current situation
=================

Currently, the Eclipse Hono lora adapter implements the AbstractVertxBasedHttpProtocolAdapter now. This means that command and control is dependent on incoming messages/connections.

To send a command, following steps need to be taken:

1.  Device is provisioned in the LNS (LoraWAN network server).
    
2.  LNS is configured to forward messages to Hono
    
3.  Device sends message, LNS forwards message to Hono (without TTD)
    
4.  Hono subscribes to commands from the business layer (this step is required to know the last gateway)
    
5.  When a command is sent to a specific device, the lorawan adapter receives the command but cannot send it to the LNS because the incoming connection is not open anymore (which is required for HTTP) and Hono does not know how to reach the LNS.
    

Proposal
========

We would like to propose a change in the lorawan adapter to enable commands to LoraWAN network servers without relying on incoming traffic. According to some investigation on our 3 most used LNSs(see further), the LNS requires an HTTP POST to the LNS to send a command to the sensor (originating from Hono in this case). The reason why we would like to implement this in Hono, is the fact that Hono is aware of which LNS is connected to which device while the business application is not aware of this. This abstraction layer is the same for incoming events/telemetry: the business application should not be aware of the network (LoraWAN, DASH7, something customer, …) to which the device is connected.

To enable Hono to send these commands to the LNS/device, hono should

*   know the LNS endpoint
    
*   be able to provide authentication (Bearer token in header or api key added to URI)
    
*   format the message
    
*   know the specific required headers per LNS
    

Al of these specifics can be implemented in the lora provders in hono. We are aware of the fact that this changes the current functionanlity a bit so that we are actually doing requests originated from hono but this is required for those LNS. We will develop this anyway because we need it for our use case and we would love to contribute this back.

LNS UseCases
============

Loriot
------

API documentation: [https://docs.loriot.io/pages/viewpage.action?pageId=7471907](https://docs.loriot.io/pages/viewpage.action?pageId=7471907)

### Send Downlink example

[https://docs.loriot.io/display/LNS/Send+Downlinks+via+API](https://docs.loriot.io/display/LNS/Send+Downlinks+via+API)

```
curl --location --request POST 'https://eu1.loriot.io/1/rest' \\
   --header 'Content-Type: application/json' \\
   --header 'Authorization: Bearer vgEADQAAAA9iY2dkdi5sb3Jpb3QuaW-ve6WTseo19IasKoGVQJSn' \\
   --header 'Content-Type: text/plain' \\
   --data-raw '{
      "cmd": "tx",
      "EUI": "BE7A000000000009",
      "port": 33,
      "confirmed": false,
      "data": "0ffff0",
      "appid": "BE01000D"
   }'
```
 

#### Response

No information available about the response whatsoever.

Firefly
-------

API documentation: [https://apidocs.fireflyiot.com/](https://apidocs.fireflyiot.com/)

### Send Downlink example

[https://apidocs.fireflyiot.com/#create-send-packet-to-device](https://apidocs.fireflyiot.com/#create-send-packet-to-device)

Sample request
POST "https://api.fireflyiot.com/api/v1/devices/eui/2564927382738492/packet?auth=secretapikey"

Sample request body
```
{
  "payload": "Test Payload",
  "encoding": "utf8",
  "port": 1,
  "confirmed": false
}
```

#### Response

Sample response
```
{
  "sent\_packet": {
    "uid": "724f0f4b-df8c-4e5b-a00d-203bdfa737ab",
    "port": 1,
    "payload": "54657374205061796C6F6164",
    "fcnt": 15
  }
}
```

No information about the return codes available

Chirpstack
----------

API documentation:

*   Raw swagger json: [https://raw.githubusercontent.com/brocaar/chirpstack-api/master/swagger/as/external/api/deviceQueue.swagger.json](https://raw.githubusercontent.com/brocaar/chirpstack-api/master/swagger/as/external/api/deviceQueue.swagger.json) (can be viewed using [https://editor.swagger.io/](https://editor.swagger.io/))
    
*   Other swaggers: [https://github.com/brocaar/chirpstack-api/tree/master/swagger/as/external/api](https://github.com/brocaar/chirpstack-api/tree/master/swagger/as/external/api)
    

### Example downlink

```
curl -X POST --header 'Content-Type: application/json' --header 'Accept: application/json' --header 'Grpc-Metadata-Authorization: Bearer <API TOKEN>' -d '{ \\ 
   "deviceQueueItem": { \\ 
     "confirmed": false, \\ 
     "data": "AQID", \\ 
     "fPort": 10 \\ 
   } \\ 
 }' 'http://localhost:8080/api/devices/0101010101010101/queue'
```
From the swagger:

![ca99e15a-1339-4ffb-9550-65728df4ff7f](https://user-images.githubusercontent.com/3315222/107958321-20c37e80-6fa2-11eb-98fb-3dc271cbead2.png)

Example payload:
```
{
  "deviceQueueItem": {
    "devEUI": "string",
    "confirmed": true,
    "fCnt": 0,
    "fPort": 0,
    "data": "string",
    "jsonObject": "string"
  }
}
```

#### Response

![38ba81ef-7851-4ab2-bac0-1efaaba1bee4](https://user-images.githubusercontent.com/3315222/107958377-33d64e80-6fa2-11eb-9455-73ee872132d8.png)

Apparently, only return code 200 is posible

**Comment by calohmn** [comment]
Thanks for the detailed proposal!

Some thoughts and questions on this:

Where exactly would the LNS endpoint and authentication information be kept in Hono?
LNS endpoint in the device registration data ("ext" section)? I guess where the authentication information should go would be something to discuss further, maybe using the device credentials data for that, although I'm not sure about that.

Would the flow for getting commands delivered to a device look something like this?
- LNS sends HTTP request for a specific device to the Lora adapter, informing the adapter about the online-state of the device, registering an adapter-internal mapping from device to associated LNS there
- the adapter sends a "[Register command consumer for device](https://www.eclipse.org/hono/docs/api/command-router/#register-command-consumer-for-device)" request to the Command Router component, mapping device to the adapter instance
- at some point later, the northbound application sends a command. It gets received by the Command Router and forwarded to the adapter (based on the previously registered mapping)
- the adapter receives the command and gets the associated LNS from the previously registered, adapter-internal mapping
- the LNS endpoint and authentication information is being determined (see question above)
- the provider type of the LNS is determined and the command is formatted according to the provider specific rules
- the adapter makes an HTTP Post request with the command to the LNS
- Possibly: if the command contains a reply-to field, the LNS response is forwarded as a command response to the northbound application
- after the device has gone offline, the LNS sends an HTTP request to the adapter, marking the device as offline, and causing the adapter to invoke "[Unregister command consumer for device](https://www.eclipse.org/hono/docs/api/command-router/#unregister-command-consumer-for-device)" in the Command Router

Instead of the explicit LNS requests to mark the device as online/offline, this way of marking the device as ready to receive commands and associating it with the LNS could also be done implicitly, when the device sends any (telemetry/event) message.
But what if commands shall be received even if the device hasn't sent a message yet and how to mark the device as offline again (maybe that's not needed, though).

**Comment by BobClaerhout** [comment]
Thanks for the response @calohmn,

I think it would be useful to configure the endpoint and the authentication information within a new section called `command`. I would propose following structure:
```
command: {
        endpoint: "https://my.lora.network.server/commands",
        headers: { < --this is a dynamic field, headers can be added at will and will be added to the outgoing request. Allows you to do basic auth or auth with a bearer token or something custom to your usecase
            Authorization: "Basic ...",
            Content - Type: "application/json",
            ...
        },
        clientCertificate: "cert", < --optional
        loraProvider: "loriot" < --optional: might be left empty so this can be reused in the http adapter(this will result in no payload mapping and using the endpoint as is.
}
```
I believe `ext` (hono docs: `Allows arbitrary properties as extension to the ones specified by the Hono API.`) is something reserved for the properties next to the Hono API and the goal of this issue is to get the command and control for lora in the hono API.

Knowing whether a device is online or offline is something very use case specific. Some devices only send messages every day, others are sending messages every 30 seconds and in some use cases they only send a message based on an external trigger. So the LNS cannot know whether a device is online or offline neither did I stumble upon a message for an upstream application from these LNSs. Therefore I would go for the implicit way you describe based on event and telemetry messages.In that case, commands cannot be send to a device which hasn't send any message yet but that's not an issue I think.

To summarize, I would add the additional functionality to the HTTP adapter and let the lora adapter override the endpoint and payload based on the given provider. If no provider is given, the payload and endpoint are not altered and will be sent as is.

One final remark: we might add a placeholder for the deviceId in the endpoint as well (i.e.: `https://my.lora.network.server/api/{devcie-id}/command`) and replace them automatically since most of the time (at least in my 3 examples) the deviceId was the only changing part in the endpoint. Makes me wonder whether the adapter still knows the deviceId when sending the message to the LNS (=gateway/lastVia)

**Comment by sophokles73** [comment]
> I think it would be useful to configure the endpoint and the authentication information within a new section called command.

Do you want to add this section to the device registration information or the tenant? I believe that most of it would actually be tenant specific, not device specific, or am I mistaken? One of the problems with this approach is that we would add secrets (tokens, private keys passwords) to this structure in clear text. That might not be an issue for some users but it might for others (most?). What's your take on that?

**Comment by BobClaerhout** [comment]
I would add it to the device registration. it's not that common but when a specific tenant has 2 different LNS in place, configuration on the tenant would not suffice. Also, the information on how to reach a gateway (LNS) simply belongs in the registration of that gateway/device IMO.

Your remark on adding secrets in plain text is indeed an issue. We might encode and decode these configurations with a given secret? The same goes for connections in Ditto as well, in that case the configuration is just stored in plain text in the database. Nevertheless, I would go for encrypting it in the database with a secret.

**Comment by sophokles73** [comment]
Well you could still define the (multiple) LNS endpoints at the tenant level once and then refer to them at the device level. That way you will need to update credentials and URLs in one place only.

**Comment by BobClaerhout** [comment]
I don't get why this is helpful. The LNS is the gateway to hono. So in case of a Lora network, you only have one gateway configured in hono. All devices are sending data through that one hono gateway and therefor you will only have one device using that defined endpoint on the tenant. I can't seem to grasp the advantage of configuring it on the tenant.

**Comment by sophokles73** [comment]
> So in case of a Lora network, you only have one gateway configured in hono. All devices are sending data through that one hono gateway and therefor you will only have one device using that defined endpoint on the tenant. I can't seem to grasp the advantage of configuring it on the tenant.

Ok, now I see. You want to attach the corresponding information to the (Hono) gateway device representing the LNS. That certainly makes sense and I agree that you would most likely have one such gateway device (per tenant) only.

@calohmn I guess a protocol adapter should be able to access a gateway device's registration information when it receives a 
command to forward to that gateway, right?
However, I also think that we will need to re-factor command processing in the protocol adapter for this to work.

**Comment by calohmn** [comment]
> I guess a protocol adapter should be able to access a gateway device's registration information when it receives a
command to forward to that gateway, right?

Yes, sure. The Lora adapter would just have to invoke `assertRegistration` on the gateway device and get the LNS data from the result.

> However, I also think that we will need to re-factor command processing in the protocol adapter for this to work.

From the first look at it, I think the refactoring will mainly need to be done in the `AbstractVertxBasedHttpProtocolAdapter` class, providing the necessary hooks/extension points for:
- registering a command consumer with unlimited lifespan (ignoring any request ttd-param) on every received telemetry/event message (if such a consumer doesn't exist yet),
- preventing that command consumer to get closed on any error during telemetry/event processing or when processing has finished,
- setting a custom handler for that command consumer which will do the Lora specific steps (get gateway/LNS device registration data, do HTTP request).

But, from looking at it a bit more closely, instead of a refactoring in the base class, I think, the Lora adapter should probably implement its own command consumer creation handling before invoking the `AbstractVertxBasedHttpProtocolAdapter#uploadTelemetryMessage` method.
Making the existing command consumer creation logic in the base class generic enough for the Lora use case doesn't look like a good idea since a central aspect there is that the command consumer is closed when the telemetry/event request handling is done.

**Comment by BobClaerhout** [comment]
@calohmn created a PR: https://github.com/eclipse/hono/issues/2482
Would be nice if you could have a quick look at it so I'm not heading in the wrong direction.

**Comment by calohmn** [comment]
@BobClaerhout Yes, I'll have a look.

**Comment by sophokles73** [comment]
@BobClaerhout can this be closed now that #2522 has been merged?

**Comment by BobClaerhout** [comment]
Yes, sure. Happy to have that in there :)
